From b57a1aec395cff5156812295c9dbaf9b7edbddcf Mon Sep 17 00:00:00 2001
From: Julius Davies <julius@apache.org>
Date: Fri, 21 Jan 2011 19:39:11 +0000
Subject: [PATCH] Fix for CODEC-101:  java.io.InputStreamReader hates it when
 InputStream.read(byte[]) returns a zero. (Adding an extra JUnit Test for good
 luck.)

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/codec/trunk@1061980 13f79535-47bb-0310-9956-ffa450edef68
---
 .../codec/binary/Base64InputStreamTest.java   | 31 +++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/src/test/org/apache/commons/codec/binary/Base64InputStreamTest.java b/src/test/org/apache/commons/codec/binary/Base64InputStreamTest.java
index 1df095d75b..5cd02fb9a2 100644
--- a/src/test/org/apache/commons/codec/binary/Base64InputStreamTest.java
+++ b/src/test/org/apache/commons/codec/binary/Base64InputStreamTest.java
@@ -17,8 +17,10 @@
 
 package org.apache.commons.codec.binary;
 
+import java.io.BufferedReader;
 import java.io.ByteArrayInputStream;
 import java.io.InputStream;
+import java.io.InputStreamReader;
 import java.util.Arrays;
 
 import junit.framework.TestCase;
@@ -64,6 +66,35 @@ public void testCodec101() throws Exception {
         assertTrue("Codec101: Second read should report end-of-stream [c=" + c + "]", c < 0);
     }
 
+    /**
+     * Another test for the CODEC-101 bug:
+     * In commons-codec-1.4 this test shows InputStreamReader explicitly hating an
+     * InputStream.read(byte[]) return of 0:
+     *
+     * java.io.IOException: Underlying input stream returned zero bytes
+     * at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:268)
+     * at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:306)
+     * at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:158)
+     * at java.io.InputStreamReader.read(InputStreamReader.java:167)
+     * at java.io.BufferedReader.fill(BufferedReader.java:136)
+     * at java.io.BufferedReader.readLine(BufferedReader.java:299)
+     * at java.io.BufferedReader.readLine(BufferedReader.java:362)
+     * at org.apache.commons.codec.binary.Base64InputStreamTest.testInputStreamReader(Base64InputStreamTest.java:75)
+     *
+     * But in commons-codec-1.5 it's fixed.  :-)
+     *
+     * @throws Exception for some failure scenarios.
+     */
+    public void testInputStreamReader() throws Exception {
+        byte[] codec101 = StringUtils.getBytesUtf8(Base64TestData.CODEC_101_MULTIPLE_OF_3);
+        ByteArrayInputStream bais = new ByteArrayInputStream(codec101);
+        Base64InputStream in = new Base64InputStream(bais);
+        InputStreamReader isr = new InputStreamReader(in);
+        BufferedReader br = new BufferedReader(isr);
+        String line = br.readLine();
+        assertNotNull("Codec101:  InputStreamReader works!", line);
+    }
+
     /**
      * Test the Base64InputStream implementation against the special NPE inducing input
      * identified in the CODEC-98 bug.
